### ISSUE TITLE
TestReverseScanWithSplitAndMerge flaky

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#2026]
The following test appears to have failed:

[#5674](https://circleci.com/gh/cockroachdb/cockroach/5674):

```
I0809 11:17:33.221491 286 kv/range_cache.go:148  adding descriptor: key="\x00\x00meta2\xff\xff" desc=range_id:4 start_key:"g" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
I0809 11:17:33.221791 286 kv/range_cache.go:269  clearing overlapping descriptor: key="\x00\x00meta2\xff\xff" desc=range_id:4 start_key:"g" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
I0809 11:17:33.224967 286 kv/range_cache.go:124  lookup range descriptor: key="\x00\x00\x00kg\x00\x01rtn-"
I0809 11:17:33.230106 286 kv/range_cache.go:148  adding descriptor: key="\x00\x00meta2\xff\xff" desc=range_id:4 start_key:"g" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
I0809 11:17:33.230544 286 kv/range_cache.go:269  clearing overlapping descriptor: key="\x00\x00meta2\xff\xff" desc=range_id:4 start_key:"g" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
--- FAIL: TestReverseScanWithSplitAndMerge (2.05s)
	dist_sender_server_test.go:306: unexpected error on ReverseScan: storage/store.go:141: end key "b" must be greater than start "c"
=== RUN TestStartEqualsEndKeyScan
I0809 11:17:33.261861 286 base/context.go:141  setting up TLS from certificates directory: test_certs
I0809 11:17:33.322542 286 base/context.go:103  setting up TLS from certificates directory: test_certs
I0809 11:17:33.380652 286 rpc/clock_offset.go:155  monitoring cluster offset
I0809 11:17:33.380944 286 multiraft/multiraft.go:446  node 100000001 starting
I0809 11:17:33.381599 286 raft/raft.go:394  100000001 became follower at term 5
I0809 11:17:33.381845 286 raft/raft.go:207  newRaft 100000001 [peers: [100000001], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0809 11:17:33.382022 286 raft/raft.go:473  100000001 is starting a new election at term 5
I0809 11:17:33.382159 286 raft/raft.go:407  100000001 became candidate at term 6
--
c119054743779948.22  127.0.0.1:0  11:17:34.871167  0           ··sending to 127.0.0.1:59124  rpc/send.go:171
c119054743779948.22  127.0.0.1:0  11:17:34.873661  0           ·reply error: *proto.Error    kv/txn_coord_sender.go:304

I0809 11:17:34.876527 286 client/db.go:450  failed AdminSplit: storage/replica_command.go:1044: cannot split range at key "\x00\x00meta2\xff\xff"
--- PASS: TestSplitByMeta2KeyMax (0.66s)
FAIL
FAIL	github.com/cockroachdb/cockroach/kv	46.373s
=== RUN TestHeartbeatSingleGroup
I0809 11:16:50.890996 299 multiraft/multiraft.go:446  node 1 starting
I0809 11:16:50.891742 299 multiraft/multiraft.go:446  node 2 starting
I0809 11:16:50.892852 299 raft/raft.go:394  1 became follower at term 5
I0809 11:16:50.893341 299 raft/raft.go:207  newRaft 1 [peers: [1,2], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0809 11:16:50.893808 299 raft/raft.go:394  2 became follower at term 5
I0809 11:16:50.894006 299 raft/raft.go:207  newRaft 2 [peers: [1,2], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0809 11:16:50.894498 299 raft/raft.go:473  1 is starting a new election at term 5
I0809 11:16:50.894911 299 raft/raft.go:407  1 became candidate at term 6
I0809 11:16:50.895320 299 raft/raft.go:456  1 received vote from 1 at term 6

```
Please assign, take a look and update the issue accordingly.

**Comment by tschottdorf** [comment]
ping @thundercw

**Comment by thundercw** [comment]
I am looking for the reason for that.

**Comment by thundercw** [comment]
@tschottdorf  The test case I has been running for more than one hour ，  but this mistake didn't happen . It may need some more specific conditions. 
```
ok  	github.com/cockroachdb/cockroach/gossip/resolver	0.013s
?   	github.com/cockroachdb/cockroach/gossip/simulation	[no test files]
ok  	github.com/cockroachdb/cockroach/keys	0.010s
ok  	github.com/cockroachdb/cockroach/kv	7.974s
ok  	github.com/cockroachdb/cockroach/multiraft	0.140s
ok  	github.com/cockroachdb/cockroach/multiraft/storagetest	3.970s
```
How about skip the test `TestReverseScanWithSplitAndMerge` first ，and then reopen it after I do more testing and fix it ?

**Comment by tschottdorf** [comment]
my hunch is that there's an error when a `ReverseScan` gets retried. I'm guessing the `ReverseScan` might've been sent to the wrong replica, and then on the next attempt the `EndKey` was lost or altered in a wrong way. Let's leave the test in for now and work on it, if it happens frequently we'll skip it.

**Comment by tschottdorf** [comment]
I can reproduce it pretty reliably (within `5s`) of

```bash
make test TESTFLAGS=-c PKG=./kv &&
while [ $? -eq 0 ]; do echo -n . && ./kv.test -test.run TestReverseScanWithSplitAndMerge --verbosity=1 &> boom; done;
less boom
```

Will take a look later.

**Comment by thundercw** [comment]
@tschottdorf  I have printed some logs in replica_command.go ,and found that the error is at the end of the function to truncate the return ranges. The inconsistent range look up may not find the new range just splited. See logs below: [b,e)was splited to[b,c) and [c,e) ,but [b,c)  was not found.
'''
I0811 05:24:58.895314 24488 storage/replica_command.go:507  [goid = 788]lookup kye ="\x00\x00meta2b", use rverse = true,method = RangeLookup
I0811 05:24:58.895518 24488 storage/replica_command.go:623  [goid = 833223656212]get reanges = [{5 "c" "e" [{1 1 1 []}] 2 []} {1 "" "b" [{1 1 1 []}] 2 []}],lookup kye ="\x00\x00meta2b", use rverse = true
I0811 05:24:58.895580 24488 storage/replica_command.go:637  [goid = 833223656212]return reanges = [{5 "c" "e" [{1 1 1 []}] 2 []} {1 "" "b" [{1 1 1 []}] 2 []}]
'''
I will modify the bug ,and take more testing. Thanks for your tools.

**Comment by tschottdorf** [comment]
I'm uploading a fix shortly.

**Comment by thundercw** [comment]
You mean you will fix this problem?

**Comment by thundercw** [comment]
I saw it.

**Comment by tschottdorf** [comment]
fixed in #2039
